pydantic/mypy.py

```diff
@@ -346,67 +426,89 @@ def transform(self) -> None:
+            return False
+        for field in fields:
+            if field.type is None:
+                return False
+
         self.add_initializer(fields, config)
-        self.add_model_construct_method(fields)
+        self.add_model_construct_method(fields, config)
         self.set_frozen(fields, frozen=config.frozen is True)
+
+        self.adjust_decorator_signatures()
+
         info.metadata[METADATA_KEY] = {
             'fields': {field.name: field.serialize() for field in fields},
             'config': config.get_values_dict(),
         }
 
-    def adjust_validator_signatures(self) -> None:
+        return True
+
+    def adjust_decorator_signatures(self) -> None:
         """When we decorate a function `f` with `pydantic.validator(...)`, `pydantic.field_validator`
         or `pydantic.serializer(...)`, mypy sees `f` as a regular method taking a `self` instance,
         even though pydantic internally wraps `f` with `classmethod` if necessary.
 
         Teach mypy this by marking any function whose outermost decorator is a `validator()`,
         `field_validator()` or `serializer()` call as a `classmethod`.
         """
-        for name, sym in self._ctx.cls.info.names.items():
+        for name, sym in self._cls.info.names.items():
             if isinstance(sym.node, Decorator):
                 first_dec = sym.node.original_decorators[0]
                 if (
                     isinstance(first_dec, CallExpr)
                     and isinstance(first_dec.callee, NameExpr)
                     and first_dec.callee.fullname in DECORATOR_FULLNAMES
                 ):
+                    # TODO: Only do this if the first argument of the decorated function is `cls`
                     sym.node.func.is_class = True
 
     def collect_config(self) -> ModelConfigData:  # noqa: C901 (ignore complexity)
         """Collects the values of the config attributes that are used by the plugin, accounting for parent classes."""
-        ctx = self._ctx
-        cls = ctx.cls
+        # ctx = self._ctx
+        cls = self._cls
         config = ModelConfigData()
 
         has_config_kwargs = False
         has_config_from_namespace = False
 
+        # Handle `class MyModel(BaseModel, <name>=<expr>, ...):`
         for name, expr in cls.keywords.items():
             config_data = self.get_config_update(name, expr)
             if config_data:
                 has_config_kwargs = True
                 config.update(config_data)
 
+        # Handle `model_config`
+        stmt: Statement | None = None
         for stmt in cls.defs.body:
             if not isinstance(stmt, (AssignmentStmt, ClassDef)):
                 continue
 
             if isinstance(stmt, AssignmentStmt):
                 lhs = stmt.lvalues[0]
-                if not isinstance(lhs, NameExpr) or lhs.name != 'model_config' or not isinstance(stmt.rvalue, CallExpr):
+                if cls.name == 'ForbidExtraModel':
+                    print('here')
```

Comment:
I know your PR is not ready for review but I am commenting as I am reviewing.

This branch does nothing.

Reply:
whoops definitely don't want this